packetbeat/protos/amqp/amqp.go

```diff
@@ -0,0 +1,883 @@
+		return nil, true
+	}
+	if data[frame.size+7] != FRAME_END_OCTET {
+		logp.Warn("Missing frame end octet in frame, discarding it")
+		return nil, true
+	}
+	frame.Type = data[0]
+	frame.channel = binary.BigEndian.Uint16(data[1:3])
+	if frame.size == 0 {
+		//frame content is nil with hearbeat frames
+		frame.content = nil
+	} else {
+		frame.content = data[7 : frame.size+7]
+	}
+	return &frame, false
+}
+
+/*
+The Method Payload, according to official doc :
+0           2           4
++----------+-----------+-------------- - -
+| class-id | method-id | arguments...
++----------+-----------+-------------- - -
+  short       short       ...
+*/
+
+func (amqp *Amqp) decodeMethodFrame(s *AmqpStream, m_data []byte) (bool, bool) {
+	class := binary.BigEndian.Uint16(m_data[0:2])
+	method := binary.BigEndian.Uint16(m_data[2:4])
+	arguments := m_data[4:]
+	s.message.Parse_arguments = amqp.Parse_arguments
+	s.message.Body_size = uint64(len(m_data[4:]))
+
+	logp.Debug("amqp", "Received frame of class %d and method %d", class, method)
+
+	if function, exists := amqp.MethodMap[class][method]; exists {
+		return function(s.message, arguments)
+	} else {
+		logp.Debug("amqpdetailed", "Received unkown or not supported method")
+		return true, false
+	}
+}
+
+/*
+Structure of a content header, according to official doc :
+0           2        4          12               14
++----------+--------+-----------+----------------+------------- - -
+| class-id | weight | body size | property flags | property list...
++----------+--------+-----------+----------------+------------- - -
+  short      short   long long        short         remainder...
+*/
+
+func (amqp *Amqp) decodeHeaderFrame(s *AmqpStream, h_data []byte) bool {
+	s.message.Body_size = binary.BigEndian.Uint64(h_data[4:12])
+	logp.Debug("amqp", "Received Header frame. A message of %d bytes is expected", s.message.Body_size)
+
+	if amqp.Parse_headers == true {
+		err := getMessageProperties(s, h_data[12:])
+		if err {
+			return false
+		}
+	}
+	return true
+}
+
+/*
+Structure of a body frame, according to official doc :
++-----------------------+ +-----------+
+| Opaque binary payload | | frame-end |
++-----------------------+ +-----------+
+*/
+
+func (s *AmqpStream) decodeBodyFrame(b_data []byte) (ok bool, complete bool) {
+	s.message.Body = append(s.message.Body, b_data...)
+
+	logp.Debug("amqp", "A body frame of %d bytes long has been transmitted",
+		len(b_data))
+	//is the message complete ? If yes, let's publish it
+	if uint64(len(s.message.Body)) < s.message.Body_size {
+		return true, false
+	} else {
+		return true, true
+	}
+}
+
+//function to get message content-type and content-encoding
+func getMessageProperties(s *AmqpStream, data []byte) bool {
+	m := s.message
+	prop1 := data[0]
+	prop2 := data[1]
+	var offset uint32 = 2
+
+	//while last bit set, we have another property flag field
+	for lastbit := 1; data[lastbit]&1 == 1; {
+		lastbit += 2
+		offset += 2
+	}
+
+	if prop1&128 == 128 {
```

Comment:
Amqp usually sets options in bits contained in a single byte. Those magic numbers are just here to check if a bit is set in the byte containing the options. There might be a more elegant solution to parse them, I agree.